<?xml version="1.0" ?>

<!-- Ths is a urdf file used to support the development of CW2 project -->
<!-- Please use xacro to clean up the URDF file, i.e using the following 4 files to represent the robot model which are easier to read and maintain-->
<!-- yourGroupidBot.xacro - A primary file that loads the following three files and contains only URDF items like joints and links -->
<!-- yourGroupidBot.gazebo â€“ It contains gazebo-specific labels including adding a Gazeo ROS control plugin  that are wrapped within gazebo -->
<!-- materials.xacro: defining all the material properties such as color -->
<!-- macros.xacro: defining macros for the caculation of inertial for example to help simplify -->

<robot name="Group1Bot">
    
  <!-- all the gazebo tags -->
  <gazebo reference="base_link_chassis">
    <material>Gazebo/Orange</material>
  </gazebo>
  <gazebo reference="left_wheel">
    <material>Gazebo/Blue</material>
  </gazebo>
  <gazebo reference="right_wheel">
    <material>Gazebo/Blue</material>
  </gazebo>

  <!--adding a differential drive plugin-->
  <gazebo>
    <plugin filename="libgazebo_ros_diff_drive.so" name="differential_drive_controller">
      <alwaysOn>true</alwaysOn>
      <updateRate>20</updateRate>
      <leftJoint>left_wheel_hinge</leftJoint>
      <rightJoint>right_wheel_hinge</rightJoint>
      <wheelSeparation>0.4</wheelSeparation>
      <wheelDiameter>0.2</wheelDiameter>
      <torque>0.1</torque>
      <commandTopic>Group1Bot/cmd_vel</commandTopic>
      <odometryTopic>odom</odometryTopic>
      <odometryFrame>odom</odometryFrame>
      <robotBaseFrame>base_link_chassis</robotBaseFrame>
      <legacyMode>true</legacyMode> 
    </plugin>
  </gazebo>

  <!--Create a robot chassis as Base Link -->
  <link name="base_link_chassis">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.5 0.3 0.07"/>
      </geometry>
      <material name="black">
        <color rgba="1.0 0.0 0.0 1.0"/>
      </material>
    </visual>

    <!-- pose and inertial -->
    <pose>0 0 0.1 0 0 0</pose>
    
    <inertial>
      <mass value="5"/>
      <origin rpy="0 0 0" xyz="0 0 0.1"/>
      <inertia ixx="0.0395416666667" ixy="0" ixz="0" iyy="0.106208333333" iyz="0" izz="0.106208333333"/>
    </inertial>
    
    <collision>
      <geometry>
        <box size="0.5 0.3 0.07"/>
      </geometry>
    </collision>
   
    <!-- caster front -->
    <visual name="caster_visual">
      <origin rpy=" 0 0 0" xyz="0.2 0 -0.05"/>
      <geometry>
        <sphere radius="0.05"/>
      </geometry>
      <material name="white">
        <color rgba="1.0 1.0 1.0 1.0"/>
      </material> 
    </visual>
    <collision name="caster_collision">
      <origin rpy=" 0 0 0" xyz="0.35 0 -0.05"/>
      <geometry>
        <sphere radius="0.05"/>
      </geometry>
    </collision>
    <inertial>
        <mass value="0.5"/>
        <inertia ixx="0.0005" ixy="0.0" ixz="0.0" iyy="0.0005" iyz="0.0" izz="0.0005"/>
      </inertial>  
  </link>
  
  <!-- Create wheel right --> 
  <link name="right_wheel">   
    <visual name="right_wheel_visual">
      <origin rpy="0 1.5707 1.5707" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.04" radius="0.1"/>
      </geometry>
      <material name="blue">
        <color rgba="0.2 0.2 1.0 1.0"/>
      </material>      
    </visual> 
    <inertial>
      <mass value="0.2"/>
      <origin rpy="0 1.5707 1.5707" xyz="0 0 0"/>
      <inertia ixx="0.00052666666" ixy="0" ixz="0" iyy="0.00052666666" iyz="0" izz="0.001"/>
    </inertial>
    
    <collision name="right_wheel_collision">
      <origin rpy="0 1.5707 1.5707" xyz="0 0 0" />
      <geometry>
        <cylinder length="0.04" radius="0.1"/>
      </geometry>
    </collision>
  </link>
    
  <!--   Joint for right wheel  -->
  <joint name="right_wheel_hinge" type="continuous">
    <origin rpy="0 0 0" xyz="-0.05 0.15 0"/>
    <child link="right_wheel" />
    <parent link="base_link_chassis"/>
    <axis rpy="0 0 0" xyz="0 1 0"/>
    <limit effort="10000" velocity="1000"/>
    <joint_properties damping="1.0" friction="1.0" />
  </joint>  
    
  <!-- Left Wheel link -->
  <link name="left_wheel">
    <visual name="left_wheel_visual">
      <origin rpy="0 1.5707 1.5707" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.04" radius="0.1"/>
      </geometry>
      <material name="blue">
        <color rgba="0.2 0.2 1.0 1.0"/>
      </material>         
    </visual>    
    <inertial>
      <mass value="0.2"/>
      <origin rpy="0 1.5707 1.5707" xyz="0 0 0"/>
      <inertia ixx="0.00052666666" ixy="0" ixz="0" iyy="0.00052666666" iyz="0" izz="0.001"/>
    </inertial>
    <collision name="left_wheel_collision">
      <origin rpy="0 1.5707 1.5707" xyz="0 0 0" />
      <geometry>
        <cylinder length="0.04" radius="0.1"/>
      </geometry>
    </collision> 
  </link>
    
  <!--   Joint for left wheel  -->
  <joint name="left_wheel_hinge" type="continuous">
    <origin rpy="0 0 0" xyz="-0.05 -0.15 0"/>
    <child link="left_wheel" />
    <parent link="base_link_chassis"/>
    <axis rpy="0 0 0" xyz="0 1 0"/>
    <limit effort="10000" velocity="1000"/>
    <joint_properties damping="1.0" friction="1.0" />
  </joint>    
    
</robot>
